### PR TITLE
Using pinched grid for generating aquifer connections

### PR DESCRIPTION
**Pull request by GitPaean**
For aquifer connections, typically/by default, the connections can only be placed on the boundary faces, which means, for these faces, the cell on one side is inside the reservoir, the cell on the other side is outside the reservoir. 

PINCH and MINPV process will deactivate the some of the cells. 

Which means, before we process the grid (PINCH, MINPV and so on), we will not be able to decide the boundary faces for the purpose of the aquifer connections. 

The purpose of this PR is to use the processed grid to generate the aquifer connections.  It is found when development numerical aquifers, but the current PR is on the setting for the existing analytical aquifers.  It improves the results for some troubling field case with analytical aquifers (generate closer numbers of the connections, but not identical yet). It does not change the results for the other field case with analytical aquifer that master branch can handle successfully. 

For numerical aquifer case, it improves the counting of the connections greatly (although there is still very small difference in the number of the connections for one of the aquifers.) .

Current understanding is that there is more logic needed for aquifers when facing non-neighboring cells (how to say whether a cell is inside or outside of the reservoirs, maybe a pinched cell should be considered still inside the reservoir? need more study here) .

Basically, it is a needed features.  

Before finishing, I would like to have some comments what is good way in handling this.  Thanks. @joakim-hove @bska 

The downsteam PRs are https://github.com/OPM/opm-grid/pull/508 , https://github.com/OPM/opm-grid/pull/508

### CODE REVIEW

**Comment by GitPaean** [comment]
Maybe I should get the first commit merged first, to the make the PR even easier.

**Comment by OPMUSER** [comment]
Note the keywords for connecting analytical and numerical aquifers are different and they have different options, have you taken that into account? Also consider writing out the AQUIFERA property in the INIT file to show where the connections are, as shown below.

image

In the figure the blue cells have a value of zero for no aquifer connection, and the connections for aquifer number one have a value of 1 (red cells). If you have five aquifers then AQUIFERA would range from zero to five. Also if there are numerical aquifers then the cells associated with them have a negative value, whereas the connections (like the analytical aquifer connections) have a positive value.

If you write out the AQUIFERA property it may save you some time in the end, rather then looking at a bunch of numbers.

Ternary plot at the end of the run for reference.

image

**Comment by GitPaean** [comment]
Thanks @OPMUSER  for the suggestion. 

> Note the keywords for connecting analytical and numerical aquifers are different and they have different options, have you taken that into account? 

I know there are some difference in the definition of the connection keywords.  The one I think might affects the number of the connection number is that for analytical aquifers, there is one item (item 9. The aquifer influx coefficient.) might combine several connections to be one. Not sure whether this is what you are referring to?

Thanks for the suggestion of keyword `AQUIFERA` in the INIT file. I will take a look.

**Comment by GitPaean** [comment]
> Thanks for the suggestion of keyword AQUIFERA in the INIT file. I will take a look.

If I am not mistaken, at the moment, we do not output this to the INIT file yet.

**Comment by bska** [comment]
> > Thanks for the suggestion of keyword AQUIFERA in the INIT file. I will take a look.
> 
> If I am not mistaken, at the moment, we do not output AQUIFERA and AQUIFERN to the INIT file yet.

That is correct.  We (Flow) currently do not support outputting any kind of aquifer information to the INIT file.  I don't know how much work it would be to add such support.  At minimum we need to actually transfer the connection data to the INIT file writing procedure and we'd need to actually characterise the distinguishing contents of the `A` and `N` arrays.

**Comment by joakim-hove** [comment]
From your PR description I understand that there is "something to discuss" with this - but I have not yet fully understood what that is. Will look more closely and reach out in the afternoon.

**Comment by OPMUSER** [comment]
@bska and @GitPaean I think it would be a good idea to implement AQUIFERA in the INIT file as it will save you a lot time in debugging the aquifer connections. The connection keywords are for the two types of aquifers are:
```
--                                                                              
--                      ANALYTIC AQUIFER CONNECTION                            
--                                                                              
--       ID     ---------- BOX ---------   CONNECT  AQF    AQF     ADJOIN         
--       NUMBER I1  I2   J1  J2   K1  K2   FACE     INFLX  MULTI   CELLS                                                                                            
AQUANCON                                                                    
         1     57  57   28  36   46  58    'I+'    1*     1*       'NO'        /                                                                            
         1    111 111   38  41   22  31    'I+'    1*     1*       'NO'        /                                                                            
         1     96  96   44  49   22  31    'I+'    1*     1*       'NO'        /                                                                         
         1     43  43   28  35   54  58    'I+'    1*     1*       'NO'        /       
         1     98  98   38  42   32  40    'I+'    1*     1*       'NO'        /       
         1     79  79   41  67    5  11    'I+'    1*     1*       'NO'        /       
         1     61  61   48  72   12  17    'I+'    1*     1*       'NO'        /
/     
--                                                                              
--       NUMERICAL AQUIFER CONNECTION                            
--                                                                              
--       ID     ---------- BOX ---------   CONNECTION  TRANS   CONNECTION            
--       NUMBER I1  I2   J1  J2   K1  K2   FACE        MULT    OPTION                                                                                       
AQUCON                                                                          
         1       1  96    1  58  28   28   'K+'        1*      1*              / 
/                                                                               
```

ADJION cells options if set to yes will allow for connections inside the reservoir grid. It is not recommended to use this option without thoroughly checking the connections in the model. The Connection option does the same thing.

So in the OPM Flow implementation are you connecting to cells "within the reservoir" as per the ADJION/CONNECTION options?

**Comment by GitPaean** [comment]
> That is correct. We (Flow) currently do not support outputting any kind of aquifer information to the INIT file. I don't know how much work it would be to add such support. At minimum we need to actually transfer the connection data to the INIT file writing procedure and we'd need to actually characterise the distinguishing contents of the A and N arrays.

Hi, @bska , it might be a good idea to have `AQUIFERA` and `AQUIFERN` to the INIT file.  We are coming back to this PR soon as other parts of the numerical aquifers are done shortly, hopefully.  If possible, could you please point me to the code I should look at about writing information to INIT file or some general ideas from you about how we should make this work?
I can check how we can manage with the Aquifer information to see whether we can create a PR for this.  Thanks.

**Comment by bska** [comment]
> If possible, could you please point me to the code I should look at about writing information to INIT file or some general ideas from you about how we should make this work?

The process of writing the INIT file is completely controlled by [`EclipseiO::writeInitial()`](https://github.com/OPM/opm-common/blob/c294e1779ad37645fa93b61e03476a7543b87889/src/opm/output/eclipse/EclipseIO.cpp#L180-L181) which does some file handling and then defers the actual file writing to [`InitiIO::write()`](https://github.com/OPM/opm-common/blob/c294e1779ad37645fa93b61e03476a7543b87889/src/opm/output/eclipse/WriteInit.cpp#L564).  Just for completeness, `EclipseIO::writeInitial()` is called from [`EclProblem::finishInit()`](https://github.com/OPM/opm-simulators/blob/ee86c9d991d4b0be23a68e33bd0a9ad771fa8fec/ebos/eclproblem.hh#L959-L961) through [`EclWriter::writeInit()`](https://github.com/OPM/opm-simulators/blob/ee86c9d991d4b0be23a68e33bd0a9ad771fa8fec/ebos/eclwriter.hh#L247).

Integer arrays like `AQUIFERA` and `AQUIFERN` don't need much extra logic apart from their values.  I don't know if there is an ordering requirement between the keywords however.  That said at minimum we would need to find a way to transfer the cells and their associate aquifer ID and aquifer type (meaning analytic vs. numeric) to the output writing layer.  Depending on how this connection information is structured on the simulator side&mdash;especially if the cells are already grouped by aquifer ID or not&mdash;I can think of at least two ways to do so.

### Data IS Grouped by Aquifer ID ###
Create a connection structure of the form
```C++
struct AquiferCellConnections
{
    std::vector<int> aquiferID;
    std::vector<std::vector<int>::size_type> cellPos;
    std::vector<int> cell;
};
```
in which the cells of `aquiferID[i]` are in `cells[cellPos[i] .. cellPos[i + 1] - 1]`.  Then we could pass an aggregate of the form
```C++
struct Aquifers {
    AquiferCellConnections analytic;
    AquiferCellConnections numeric;
};
```
as an argument to `EclipseIO::writeInitial`.

### Data Is NOT Grouped by Aquifer ID ###
I'd choose a structure similar to the one above, but represent the `AquiferCellConnections` along the lines of
```C++
struct AquiferCellConnection
{
    int aquiferID;
    int cell;
};

struct AquiferCellConnections
{
    std::vector<AquiferCellConnection> connection;
};
```

On a different note, I expect to do some work on writing aquifer information to the restart files (`*.UNRST`) soon so maybe we can coordinate this effort?

**Comment by GitPaean** [comment]
Thanks @bska for the explanation. I have limited working capacity today, will get back to you hopefully tomorrow.

**Comment by GitPaean** [comment]
> That said at minimum we would need to find a way to transfer the cells and their associate aquifer ID and aquifer type (meaning analytic vs. numeric) to the output writing layer. Depending on how this connection information is structured on the simulator side

This information might be complete in `AquiferConfig` in `EclipseState`,  especially if this PR going in.  We can convert the information from `AquiferConfig` to one of the forms you mentioned below. 

> On a different note, I expect to do some work on writing aquifer information to the restart files (*.UNRST) soon so maybe we can coordinate this effort?

We can do that. I will try to draft an implementation for the `AQUIFERA` and `AQUIFERN`, then we can have a discussion.

**Comment by bska** [comment]
> > That said at minimum we would need to find a way to transfer the cells and their associate aquifer ID and aquifer type (meaning analytic vs. numeric) to the output writing layer. Depending on how this connection information is structured on the simulator side
> 
> This information might be complete in `AquiferConfig` in `EclipseState`, especially if this PR going in. We can convert the information from `AquiferConfig` to one of the forms you mentioned below.

If the information is already in `AquiferConfig` then we won't need to do any of the translations.  Then we'll be able to just create the output arrays directly.

**Comment by GitPaean** [comment]
> If the information is already in AquiferConfig then we won't need to do any of the translations. Then we'll be able to just create the output arrays directly.

All the field props in the INIT file, looks like corresponding to the active cells except `PORV` (which looks like has the final say about the active or not). 

`AQUIFERA` and `AQUIFERN` are also corresponding to the active cells.  Similar to the purpose of this PR, at the end, we might need the information of the processed grid to finalize the output of these two keywords.   Anyway, let us have a version based on `AquiferConfig` and `EclipseGrid` first.

**Comment by GitPaean** [comment]
After https://github.com/OPM/opm-simulators/pull/3039 goes in, this PR will be the next to address. Very likely, it will be one of the last pieces to make the aquifer models complete. But this part is also deep in the grid processing, some help will be requested during the development from the people familiar with the grid processing code.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
There are many functions in this PR which get a new argument `std::vector<int>& actnum` in addition to the grid; but it seems the actnum vector is only:
```C++
const auto& actnum = grid.getACTNUM();
```
i.e. the actnum argument seems a bit superfluous? Am I misunderstanding something - or maybe there is a plan that the actnum argument will in the future be different from the actnum information embedded in the grid.

File: src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquifers.cpp [hunk 1]
```diff
@@ -161,11 +161,11 @@ namespace Opm {
...
-        std::vector<NNCdata> nncs;
-        for ([[maybe_unused]] const auto& [id, aquifer] : this->m_aquifers) {
-            auto aqu_nncs = aquifer.aquiferNNCs(grid, fp);
-            nncs.insert(nncs.end(), aqu_nncs.begin(), aqu_nncs.end());
-        }
+        // begin with the NNCs from the cells
+        auto nncs = this->aquiferCellNNCs();
+        // appending the NNCs from the connections
+        auto con_nncs = this->aquiferConnectionNNCs(grid, fp);
+        nncs.insert(nncs.end(), con_nncs.begin(), con_nncs.end());
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquifers.cpp:168]: Here you can move out of `con_nncs` with a move iterator:
```
nncs.insert( nncs.end(), std::make_move_iterator(con_nncs.begin()), std::make_move_iterator(con_nns.end()));
```

File: opm/parser/eclipse/EclipseState/Aquifer/AquiferConfig.hpp [hunk 2]
```diff
@@ -53,6 +53,7 @@ class AquiferConfig {
     bool hasNumericalAquifer() const;
     bool hasAnalyticalAquifer() const;
     const NumericalAquifers& numericalAquifers() const;
+    NumericalAquifers& mutableNumericalAquifers() const;
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Aquifer/AquiferConfig.hpp:56]: Hehe - thank you for naming it `mutable`; that made the intent screaming clear. Where will the callsite actually modifying this object be? 

I must admit my initial reaction to exporting a mutable object in this way is quite skeptical, but we'll see. I realize that due to the dance up and down between common &leftrightarrow; grid &leftrightarrow; simulators it might be difficult to avoid?

**Comment by GitPaean** [comment]
Sorry, this PR is not ready for review. It is the PR https://github.com/OPM/opm-common/pull/2389 ready for review for numerical aquifers. @joakim-hove . This PR will be updated later to handle analytical aquifers.

**Comment by GitPaean** [comment]
Sorry for the confusion. Splitting PRs was a good idea, but I see it causes confusion here too.

**Comment by blattms** [comment]
@GitPaean Shall we close this?

**Comment by GitPaean** [comment]
>  @GitPaean Shall we close this?

I think so. We can revisit those PRs when coming to this.

**Comment by blattms** [comment]
Thanks.
Just a comment: PINCH/MINPV processing is not done in opm-common, but during grid processing in opm-grid.
